From: Zheng Liu <wenqing.lz@taobao.com>
Subject: [PATCH v5 7/8] ext4: request accouting in direct io
Patch-mainline: in house patch

ext4_ios_submit_io() function is defined to be called by __blockdev_direct_IO()
to count direct io.

Signed-off-by: Wang Shaoyan <wangshaoyan.pt@taobao.com>
Signed-off-by: Zheng Liu <wenqing.lz@taobao.com>

Index: linux-2.6.32-220.4.2.el5/fs/ext4/ext4.h
===================================================================
--- linux-2.6.32-220.4.2.el5.orig/fs/ext4/ext4.h
+++ linux-2.6.32-220.4.2.el5/fs/ext4/ext4.h
@@ -1816,6 +1816,8 @@ extern qsize_t *ext4_get_reserved_space(
 extern int flush_aio_dio_completed_IO(struct inode *inode);
 extern void ext4_da_update_reserve_space(struct inode *inode,
 					int used, int quota_claim);
+extern void ext4_ios_submit_io(int rw, struct bio *bio,
+			       struct inode *inode, loff_t file_offset);
 /* ioctl.c */
 extern long ext4_ioctl(struct file *, unsigned int, unsigned long);
 extern long ext4_compat_ioctl(struct file *, unsigned int, unsigned long);
Index: linux-2.6.32-220.4.2.el5/fs/ext4/inode.c
===================================================================
--- linux-2.6.32-220.4.2.el5.orig/fs/ext4/inode.c
+++ linux-2.6.32-220.4.2.el5/fs/ext4/inode.c
@@ -3842,9 +3842,10 @@ static ssize_t ext4_ind_direct_IO(int rw
 	}
 
 retry:
-	ret = blockdev_direct_IO(rw, iocb, inode, inode->i_sb->s_bdev, iov,
-				 offset, nr_segs,
-				 ext4_get_block, NULL);
+	ret = __blockdev_direct_IO(rw, iocb, inode, inode->i_sb->s_bdev,
+				   iov, offset, nr_segs,
+				   ext4_get_block, NULL, ext4_ios_submit_io,
+				   DIO_LOCKING | DIO_SKIP_HOLES);
 	if (ret == -ENOSPC && ext4_should_retry_alloc(inode->i_sb, &retries))
 		goto retry;
 
@@ -4235,11 +4236,11 @@ static ssize_t ext4_ext_direct_IO(int rw
 			EXT4_I(inode)->cur_aio_dio = iocb->private;
 		}
 
-		ret = blockdev_direct_IO(rw, iocb, inode,
-					 inode->i_sb->s_bdev, iov,
-					 offset, nr_segs,
-					 ext4_get_block_dio_write,
-					 ext4_end_io_dio);
+		ret = __blockdev_direct_IO(rw, iocb, inode, inode->i_sb->s_bdev,
+					   iov, offset, nr_segs,
+					   ext4_get_block_dio_write,
+					   ext4_end_io_dio, ext4_ios_submit_io,
+					   DIO_LOCKING | DIO_SKIP_HOLES);
 		if (iocb->private)
 			EXT4_I(inode)->cur_aio_dio = NULL;
 		/*
@@ -4279,6 +4280,18 @@ static ssize_t ext4_ext_direct_IO(int rw
 	return ext4_ind_direct_IO(rw, iocb, iov, offset, nr_segs);
 }
 
+void ext4_ios_submit_io(int rw, struct bio *bio,
+			struct inode *inode, loff_t file_offset)
+{
+	int tmprw = !!rw;
+
+	/* ext4 io type accouting */
+	__ext4_io_stat(inode->i_sb, tmprw, EXT4_IOS_REGULAR_DATA,
+		       ((bio->bi_size + PAGE_CACHE_SIZE - 1)
+			>> PAGE_CACHE_SHIFT));
+	submit_bio(rw, bio);
+}
+
 static ssize_t ext4_direct_IO(int rw, struct kiocb *iocb,
 			      const struct iovec *iov, loff_t offset,
 			      unsigned long nr_segs)
